RULES/hdl/questa.go

```diff
@@ -101,456 +101,524 @@ var rules = make(map[string]bool)
-      log_file.String(), rule.Lib(), top, target)
-
-    // Set up parameters
-    if param_set != "" {
-      // Check that the parameters exist
-      if params, ok := rule.Params[param_set]; ok {
-        // Add parameters for all generics
-        for param, value := range params {
-          cmd = fmt.Sprintf("%s -G %s=%s", cmd, param, value)
-        }
-      }
-    }
-
-    if rule.TestCaseGenerator != nil {
-      deps = append(deps, rule.TestCaseGenerator)
-    }
-
-    // Add the rule to run 'vopt'.
-    ctx.AddBuildStep(core.BuildStep{
-      Out:   log_file,
-      Ins:   deps,
-      Cmd:   cmd,
-      Descr: fmt.Sprintf("vopt: %s %s", rule.Lib()+"."+top, target),
-    })
-
-    // Note that we created this rule
-    rules[log_file.String()] = true
-  }
+	top := "board"
+	if rule.Top != "" {
+		top = rule.Top
+	}
+
+	cover_flag := ""
+	log_file_suffix := "vopt.log"
+	if Coverage.Value() {
+		cover_flag = "+cover"
+		log_file_suffix = "vopt_cover.log"
+	}
+
+	log_files := []core.OutPath{}
+	targets := []string{}
+	params := []string{}
+	if rule.Params != nil {
+		for key, _ := range rule.Params {
+			log_files = append(log_files, rule.Path().WithSuffix("/"+key+"_"+log_file_suffix))
+			targets = append(targets, key+"_"+rule.Target())
+			params = append(params, key)
+		}
+	} else {
+		log_files = append(log_files, rule.Path().WithSuffix("/"+log_file_suffix))
+		targets = append(targets, rule.Target())
+		params = append(params, "")
+	}
+
+	for i := range log_files {
+		log_file := log_files[i]
+		target := targets[i]
+		param_set := params[i]
+
+		// Skip if we already have a rule
+		if rules[log_file.String()] {
+			return
+		}
+
+		cmd := fmt.Sprintf("vopt %s %s +acc=%s -l %s -work %s %s -o %s",
+			common_flags, cover_flag, Access.Value(),
+			log_file.String(), rule.Lib(), top, target)
+
+		// Set up parameters
+		if param_set != "" {
+			// Check that the parameters exist
+			if params, ok := rule.Params[param_set]; ok {
+				// Add parameters for all generics
+				for param, value := range params {
+					cmd = fmt.Sprintf("%s -G %s=%s", cmd, param, value)
+				}
+			}
+		}
+
+		if rule.TestCaseGenerator != nil {
+			deps = append(deps, rule.TestCaseGenerator)
+		}
+
+		// Add the rule to run 'vopt'.
+		ctx.AddBuildStep(core.BuildStep{
+			Out:   log_file,
+			Ins:   deps,
+			Cmd:   cmd,
+			Descr: fmt.Sprintf("vopt: %s %s", rule.Lib()+"."+top, target),
+		})
+
+		// Note that we created this rule
+		rules[log_file.String()] = true
+	}
+}
+
+// Create a simulation script
+func doFile(ctx core.Context, rule Simulation) {
+	// Do-file script
+	params := DoFileParams{
+		Lib: rule.Lib(),
+	}
+
+	if rule.WaveformInit != nil {
+		params.WaveformInit = rule.WaveformInit.String()
+	}
+
+	doFile := rule.Path().WithSuffix("/" + "vsim.do")
+	ctx.AddBuildStep(core.BuildStep{
+		Out:   doFile,
+		Data:  core.CompileTemplate(do_file_template, "do", params),
+		Descr: fmt.Sprintf("vsim: %s", doFile.Relative()),
+	})
```

Comment:
You could just not use the DBT build step method here, and instead call the template functions.
It would recompile the doFile every time but that shouldn't hurt performance really.
Maybe then you could call it in QuestaCmd instead of during the build phase.

Reply:
I really wanted to avoid having DBT rebuild the Ninja build file all the time. I just don't think it's very scalable in the long run. It's also more difficult to debug as things tend to move around a bit in the build file depending on where you execute the `dbt` command from.